### ISSUE TITLE
[$250] [Report Fields] Title - Form fields in expense are updated when the formula changes, only when the page refreshes

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#37567]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.45-3
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - internal Team
**Slack conversation:**

**Issue found when executing PR** https://github.com/Expensify/App/pull/37160
## Action Performed:
Preconditions: be a member of a collect policy, have no active expense report with this policy, and have the "report fields" beta turned on.
1. Login as a member in NewDot - > Request money from the workspace
2. Login as admin in OldDot -> Change the formula for an expense report (e.g. {report:type})
3. Check a previously created expense, and verify the fields are populated correctly with the optimistic data

## Expected Result:
The expense report formula is updated automatically

## Actual Result:
The expense report formula is updated only after refreshing the page

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/77bbb947-a8c7-4204-b540-8730131230de

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0110c6cbcb96066337</li>
        <li>Upwork Job ID: 1772732351929339904</li>
        <li>Last Price Increase: 2024-04-30</li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
@strepanier03  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by strepanier03** [comment]
I'm going to raise this internally because this behavior mirrors what happens on olddot right now. The formula for the report title when updated isn't reflected on existing reports until the report itself is updated in some way. 

I'll update here shortly.

**Comment by strepanier03** [comment]
Raised [here](https://expensify.slack.com/archives/C03U7DCU4/p1709681230148829).

Perhaps @Expensify/design knows?

**Comment by strepanier03** [comment]
Held on https://github.com/Expensify/App/issues/36170

**Comment by shawnborton** [comment]
Hmm I have no idea if we would expect report title formula changes made on OldDot to automatically push to NewDot without needing any kind of page refresh. Tapping my buddy @trjExpensify for a second opinion, otherwise I say we loop in engineering.

**Comment by trjExpensify** [comment]
Yeah, I think broadly where we can, we should make OldDot/NewDot changes update automatically cross-platform -- but there's for'sure a lot of places that's not happening.

Here, I'd only expect `open` reports to update though (I'd have said `processing` as well for instant, but looks like we're cutting report titles from that setup) -- else we get into the territory of updating the report titles of a bunch of historical reports that have already been `approved, `reimbursed` or `closed` losing whatever reference they had at the time of taking those actions.

As for whether this is should be working or not right now.. @thienlnam @paultsimura the issue was created when executing the testing steps of [#37160](https://github.com/Expensify/App/pull/37160), was it supposed to be added there or is this an adjacent consideration?

**Comment by thienlnam** [comment]
> Hmm I have no idea if we would expect report title formula changes made on OldDot to automatically push to NewDot without needing any kind of page refresh. Tapping my buddy @trjExpensify for a second opinion, otherwise I say we loop in engineering.

This is known to not work right now, but should work in the future. We're currently working on a migration of some data storage in the BE in this [issue](https://github.com/Expensify/App/issues/36170)

**Comment by trjExpensify** [comment]
Perfect!

**Comment by trjExpensify** [comment]
Added this to #wave-control as report fields will be available there ultimately.

**Comment by paultsimura** [comment]
@trjExpensify @thienlnam can you reproduce it at all?
For me, the App crashes (in Prod as well) when I change the formula in OD (reported [here](https://expensify.slack.com/archives/C049HHMV9SM/p1709755227423089)).

**Comment by thienlnam** [comment]
It's probably related to this [issue](https://github.com/Expensify/App/issues/36170) as well - reportFields are behind beta so we should be okay

**Comment by strepanier03** [comment]
Assigned to a project and priority. Holding.

**Comment by JmillsExpensify** [comment]
Now that the PR for #36170 has hit staging, can we remove the hold from this issue?

**Comment by strepanier03** [comment]
Yup, thanks for the heads up @JmillsExpensify - I'll work on this today.

**Comment by Pujan92** [comment]
Thanks @nexarvo for the proposal. Your RCA and suggestion seems on the right track where we need to make an openReport API call to fetch latest report name. But your proposed code snippet will make multiple unnecessary openReport API calls, could you plz consider this issue and update your proposal.

**Comment by Pujan92** [comment]
@nexarvo [proposal](https://github.com/Expensify/App/issues/37567#issuecomment-2021598744) looks good and fixes the issue. However, I am seeing 2 `openReport` api calls on formula change with the proposed solution. It seems due to the pusher event fired with the key `text_title` and `openReport` providing `expensify_text_title`, there is some inconsistency in the names which the internal engineer can help to solve.

<details>
<summary>Pusher Event(text_title)</summary>

[info] [Report] Handled multipleEvents event sent by Pusher - {"updates":[{"data":[{"key":"policy_4129F81616940287","onyxMethod":"merge","value":{"fieldList":{**"text_title"**:{"defaultExternalID":null,"defaultValue":"{report:type} {report:startdate} {report:policyname}","deletable":false,"disabledOptions":[],"externalID":null,"externalIDs":[],"fieldID":"text_title","isTax":false,"keys":[],"name":"title","orderWeight":0,"origin":null,"target":"expense","type":"formula","value":null,"values":[]}}}}],"eventType":"onyxApiUpdate"}],"lastUpdateID":226735216,"previousUpdateID":226733699}

</details>

<details>
<summary>OpenReport API(expensify_text_title)</summary>

<img width="1351" alt="Screenshot 2024-04-01 at 10 13 23" src="https://github.com/Expensify/App/assets/14358475/dc472ce9-2ec0-436c-a537-78fa59cd661e">

</details>

🎀👀🎀 C+ reviewed

**Comment by youssef-lr** [comment]
I'm not sure the proposed solution will fully fix this issue. What happens if we have multiple reports in the LHN? In this case only the one selected will be update in the report header in the LHN, the rest of the reports will keep old names and this would look like a bug to the user. If we are storing the report field in the policy, why not generate it in the frontend everytime it changes [here](https://github.com/Expensify/App/blob/b25eb15b7122d093c60bf77b8c363187903f1248/src/libs/ReportUtils.ts#L2173) instead of returning `report.reportName`?

**Comment by thienlnam** [comment]
>  It seems due to the pusher event fired with the key text_title and openReport providing expensify_text_title, there is some inconsistency in the names which the internal engineer can help to solve.

I have a [PR](https://github.com/Expensify/Auth/pull/10414) up that should fix this, and only return text_title (this is what OldDot is using)

**Comment by Pujan92** [comment]
Thanks @youssef-lr for the review, Indeed selected proposal will solve the issue only for the opened report and won't update all LHN data. 

> If we are storing the report fields in the policy, why not generate the naw name in the frontend everytime the formula changes [here](https://github.com/Expensify/App/blob/b25eb15b7122d093c60bf77b8c363187903f1248/src/libs/ReportUtils.ts#L2173) instead of returning report.reportName?

This sounds good to me and we can take advantage of util `populateOptimisticReportFormula` which generates the report name with the formula. 

With that, This issue is still open to proposals.

**Comment by Pujan92** [comment]
Thanks @paultsimura for the info :). Also, could you plz share the link of that discussion?

**Comment by youssef-lr** [comment]
Still waiting for proposals

**Comment by thienlnam** [comment]
I wonder if this just needs to be resolved in the BE - when you update the formula, the reportName on the title enforced report should also get updated with the new reportName. This already happens in OldDot - so I wonder if we're maybe not sending out a pusher update for all of the impacted expense reports on that workspace

**Comment by youssef-lr** [comment]
>so I wonder if we're maybe not sending out a pusher update for all of the impacted expense reports on that workspace

Wouldn't that be a heavy update? I'm curious for your thoughts about generating the name in the frontend based on the formula, is there any issue with this approach?

**Comment by thienlnam** [comment]
Potentially, though only reports that are open / processing would get updated - as anything that is closed / approved does not get changed.

We could have the name generated in the FE based on the formula, the only concern is that it's not really the source of truth and doesn't have all the same formula calculations as the BE does. Maybe that's not a concern though - and we could only for open/processing reports where the formula is enforced generate an optimistic approach for a report title, and for everything else we use report.reportName

**Comment by youssef-lr** [comment]
> though only reports that are open / processing would get updated 

ahh, in this case I think a pusher update is more straightforward and requires less changes

**Comment by strepanier03** [comment]
@youssef-lr @thienlnam - I'm coming back from being OoO and I'm not sure I follow the next steps based on the conversation above.

Is the next step to change to Internal or are we waiting for proposals still?

**Comment by youssef-lr** [comment]
I think we decided on fixing this in the backend? @thienlnam is that right?

**Comment by thienlnam** [comment]
Yup that's correct - I won't be able to get around to this until after may 1st, but happy to help review if you're interested in taking it on

**Comment by strepanier03** [comment]
Thank you to you both! I'll continue to follow along until I can take action.

**Comment by youssef-lr** [comment]
No update yet, this can wait til after May 1st.

**Comment by youssef-lr** [comment]
Will make this a weekly until May 1st.

**Comment by Pujan92** [comment]
We can switch label to internal here.

**Comment by youssef-lr** [comment]
No update yet.

**Comment by JmillsExpensify** [comment]
@youssef-lr What do you think about circling back on this one? I know you're still working on delegated splits, though this is one of two remaining polish issues for report fields.

**Comment by youssef-lr** [comment]
I think I can get to this next week.

**Comment by strepanier03** [comment]
I'm going to be out of the office all of next week, June 4-12th. If you need BZ action during that time please post in #bug-zero for a hand.

**Comment by youssef-lr** [comment]
No update yet, I'll try next week.

**Comment by youssef-lr** [comment]
I'll get started on this on Friday

**Comment by youssef-lr** [comment]
I haven't managed to get to this one due to working on stuff tied to deadlines or fast-apis

**Comment by strepanier03** [comment]
Thanks @youssef-lr.

**Comment by strepanier03** [comment]
Likely still not a priority, moving to monthly until it is.

**Comment by strepanier03** [comment]
I'm going to close this out for a few different reasons:

- The repro steps aren't possible to follow as we've locked Report Fields behind the Control paywall now. 
- I cannot repro this any longer even for a Control workspace. The report fields are updating in real-time on the expense, including adding and removing fields.
